src/index.js

```diff
@@ -1,135 +1,216 @@
+function addCompilationError(compilation, message) {
+  compilation.errors.push(`Sentry CLI Plugin: ${message}`);
+}
+
+class SentryCliPlugin {
+  constructor(options = {}) {
+    // By default we want that rewrite is true
+    this.options = Object.assign({ rewrite: true }, options);
+    this.options.include = toArray(options.include);
+    this.options.ignore = toArray(options.ignore);
+
+    this.cli = this.getSentryCli();
+    this.release = this.getReleasePromise();
   }
 
-  if (typeof originalEntry === 'string') {
-    return [newEntry, originalEntry];
+  /** Returns whether this plugin is in dryRun mode. */
+  isDryRun() {
+    return this.options.dryRun === true;
   }
 
-  if (typeof originalEntry === 'function') {
-    return () =>
-      Promise.resolve(originalEntry()).then(entry => injectEntry(entry, newEntry));
+  /** Creates a new Sentry CLI instance. */
+  getSentryCli() {
+    if (this.isDryRun()) {
+      return {
+        releases: {
+          proposeVersion: () => Promise.resolve('1.0.0-dev'),
+          new: () => Promise.resolve(),
+          uploadSourceMaps: () => Promise.resolve(),
+          finalize: () => Promise.resolve(),
+        },
+      };
+    }
+
+    return new SentryCli(this.options.configFile);
   }
 
-  return newEntry;
-}
+  /**
+   * Returns a Promise that will solve to the configured release.
+   *
+   * If no release is specified, it uses Sentry CLI to propose a version. The
+   * release string is always trimmed.
+   */
+  getReleasePromise() {
+    return (this.options.release
+      ? Promise.resolve(this.options.release)
+      : this.cli.releases.proposeVersion()
+    ).then(version => `${version}`.trim());
+  }
+
+  /** Checks if the given named entry point should be handled. */
+  checkEntry(key) {
+    const { entries } = this.options;
+    if (!entries) {
+      return true;
+    }
+
+    if (typeof entries === 'function') {
+      return entries(key);
+    }
 
-function injectRelease(compiler, versionPromise) {
-  if (typeof compiler.options === 'undefined') {
-    // Gatekeeper because we are not running in webpack env
-    // probably just tests
-    return;
+    if (entries instanceof RegExp) {
+      return entries.test(key);
+    }
+
+    if (Array.isArray(entries)) {
+      return entries.indexOf(key) >= 0;
+    }
+
+    throw new Error(
+      'Invalid `entries` option: Must be an array, RegExp or function'
+    );
   }
-  const changedCompiler = compiler;
-  changedCompiler.options.entry = injectEntry(
-    changedCompiler.options.entry,
-    path.join(__dirname, 'sentry-webpack.module.js')
-  );
-  if (typeof changedCompiler.options.module === 'undefined') {
-    changedCompiler.options.module = {};
+
+  /** Injects the release string into the given entry point. */
+  injectEntry(originalEntry, newEntry) {
+    if (Array.isArray(originalEntry)) {
+      return [newEntry].concat(originalEntry);
+    }
+
+    if (originalEntry !== null && typeof originalEntry === 'object') {
```

Comment:
I know it's nitpicking, but `originalEntry = ['someEntry']` will pass this check as well, and I'm pretty sure we just want plain objects

Reply:
Can you clarify please?

Reply:
If `originalEntry` is an array (I'm not even sure if it's possible at this point of the code, so you can just ignore me if it's not), then it'll still execute this code, as `(originalEntry !== null && typeof originalEntry === 'object')` check doesn't prevent it from doing so.

```js
var originalEntry = ['foo'];
console.log(originalEntry !== null && typeof originalEntry === 'object'); // true
```

Which mean that you'll do:

```js
Object.keys(originalEntry).forEach(key => {
  ...
});
```

for every single index, as `Object.keys(array)` returns an array of available indexes, eg. `Object.keys(['a', 'b', 'c']) => [0, 1, 2]`

Reply:
Ah okay. So if originalEntry is an array, it will be handled by the branch above already. Thanks for the update!

Reply:
Right... 🤦‍♂️